WalletWasabi/WabiSabi/Models/DecompositionAlgs/RestrictedSetDecomposer.cs

```diff
@@ -0,0 +1,56 @@
+using NBitcoin;
+using System.Linq;
+
+namespace WalletWasabi.WabiSabi.Models.DecompositionAlgs
+{
+	public class RestrictedSetDecomposer
+	{
+		private Decomposition Decomposition { get; } = new();
+
+		public RestrictedSetDecomposer()
+		{
+			Decomposition.Extend(new Coin() { Amount = Money.Zero });
+		}
+
+		public virtual Money? PickNextAmount(Money remaining, Money dustThreshold)
+		{
+			return LargestDenomBelowIncl(remaining)?.Amount;
+		}
+
+		public void Decompose(Money amount, Money dustThreshold, FeeRate feeRate)
+		{
+			Money remaining = amount - feeRate.GetFee(Coin.OutputVbytes);
+
+			while (remaining > dustThreshold)
+			{
+				Money? denom = PickNextAmount(remaining, dustThreshold);
+				if (denom is null || denom == Money.Zero)
+				{
+					break;
+				}
+				Coin coin = new() { Amount = denom };
+				Decomposition.Extend(coin);
+				remaining -= coin.EffectiveCost(feeRate);
+			}
+
+			if (remaining > dustThreshold)
+			{
+				Coin coin = new() { Amount = remaining };
+				Decomposition.Extend(coin);
+			}
+		}
+
+		private Coin? LargestDenomBelowIncl(Money amount)
```

Comment:
It seems it can never return `null`

Reply:
The method looks like a `Decomposition.First(x => x.Amount < amount)`. I am  not sure if this should be a `First()` or a `FirstOrDefault`

Reply:
> It seems it can never return `null`

it may if the amount is less than the smallest value in the set

Reply:
The smallest amount in the set is zero. There won't be any smaller value than that.

Reply:
Zero is added by default as far as I can see. 

```
@attr.s(auto_attribs=True, kw_only=True, slots=True)
class RestrictedSetDecomposer(object):
    def _normalize_value_set(values: Iterable[int]) -> Tuple[int]:
        return tuple(sorted(set(chain([0], values))))

    # we keep the values as an ordered list
    values : Tuple[int] = attr.ib(converter=_normalize_value_set)
```

Reply:
I refactored the whole stuff to be functional.